pkg/envoy/registry/registry.go

```diff
@@ -10,43 +10,66 @@ func NewProxyRegistry(mapper ProxyServiceMapper, msgBroker *messaging.Broker) *P
 	return &ProxyRegistry{
 		ProxyServiceMapper: mapper,
 		msgBroker:          msgBroker,
+		connectedProxies:   make(map[string]*envoy.Proxy),
 	}
 }
 
 // RegisterProxy registers a newly connected proxy.
 func (pr *ProxyRegistry) RegisterProxy(proxy *envoy.Proxy) {
-	pr.connectedProxies.Store(proxy.UUID.String(), proxy)
+	uuid := proxy.UUID.String()
+	if pr.GetConnectedProxy(uuid) != nil {
+		log.Debug().Str("proxy", proxy.String()).Msgf("Proxy %s already registered", proxy.String())
+		return
+	}
+
+	pr.mu.Lock()
+	pr.connectedProxies[uuid] = proxy
+	pr.mu.Unlock()
 	log.Debug().Str("proxy", proxy.String()).Msg("Registered new proxy")
 }
 
 // GetConnectedProxy loads a connected proxy from the registry.
 func (pr *ProxyRegistry) GetConnectedProxy(uuid string) *envoy.Proxy {
-	p, ok := pr.connectedProxies.Load(uuid)
+	pr.mu.Lock()
+	p, ok := pr.connectedProxies[uuid]
+	pr.mu.Unlock()
+
 	if !ok {
 		return nil
 	}
-	return p.(*envoy.Proxy)
+	return p
 }
 
 // UnregisterProxy unregisters the given proxy from the catalog.
-func (pr *ProxyRegistry) UnregisterProxy(p *envoy.Proxy) {
-	pr.connectedProxies.Delete(p.UUID.String())
-	log.Debug().Msgf("Unregistered proxy %s", p.String())
+func (pr *ProxyRegistry) UnregisterProxy(proxy *envoy.Proxy) {
+	uuid := proxy.UUID.String()
+	if pr.GetConnectedProxy(uuid) == nil {
```

Comment:
nit: this log doesn't add much value IMO, can just simply call delete within the lock